common/achievement/evaluate.ts

```diff
@@ -0,0 +1,117 @@
+
+export async function evaluateAchievement(condition: string, dataAggregation: ConditionDataAggregations, metrics: string[]): Promise<EvaluationResult> {
+    const achievementEvents = await prisma.achievement_event.findMany({ where: { metric: { in: metrics } } });
+
+    const eventsByMetric: Record<string, Achievement_event[]> = {}; // Store events per metric
+    for (const event of achievementEvents) {
+        if (!eventsByMetric[event.metric]) {
+            eventsByMetric[event.metric] = [];
+        }
+        eventsByMetric[event.metric].push(event);
+    }
+
+    const resultObject: Record<string, number | string | boolean> = {};
+
+    for (const key in dataAggregation) {
+        if (!dataAggregation[key]) {
+            return;
+        }
+        const dataAggregationObject = dataAggregation[key];
+        const metricName = dataAggregationObject.metric;
+
+        const bucketCreator = dataAggregationObject.createBuckets || 'default';
+        const bucketAggregator = dataAggregationObject.bucketAggregator || 'count';
+
+        const aggregator = dataAggregationObject.aggregator;
+
+        const eventsForMetric = eventsByMetric[metricName];
+        if (!eventsForMetric) {
+            return;
+        }
+        // we take the relation from the first event, that posesses one, in order to create buckets from it, if needed
+        const relation = eventsForMetric.find((event) => event.relation)?.relation;
+
+        const bucketCreatorFunction = bucketCreatorDefs[bucketCreator].function;
+        const bucketAggregatorFunction = aggregators[bucketAggregator].function;
+        const aggFunction = aggregators[aggregator].function;
+
+        if (!bucketCreatorFunction || !bucketAggregatorFunction || !aggFunction) {
+            return;
+        }
+
+        const buckets = await bucketCreatorFunction(relation);
+        const bucketEvents = createBucketEvents(eventsForMetric, buckets);
+
+        const bucketAggr = bucketEvents.map(
+            (bucketEvent): BucketEventsWithAggr => ({
+                ...bucketEvent,
+                aggregation: bucketAggregatorFunction(bucketEvent.events.map((event) => event.value)),
+            })
+        );
+
+        const valuesFromBucketAggr = bucketAggr.map((bucket) => bucket.aggregation);
+
+        const value = aggFunction(valuesFromBucketAggr);
+        resultObject[key] = value;
+    }
+
+    const evaluate = swan.parse(condition);
+    const value: boolean = await evaluate(resultObject);
+
+    return {
+        conditionIsMet: value,
+        resultObject,
+    };
+}
+
+export function createBucketEvents(events: Achievement_event[], bucketConfig: BucketConfig): BucketEvents[] {
+    switch (bucketConfig.bucketKind) {
+        case 'default':
+            return createDefaultBuckets(events, bucketConfig);
+        case 'time':
+            return createTimeBuckets(events, bucketConfig);
+        case 'filter':
+            return createFilterBuckets(events, bucketConfig);
```

Comment:
I still don't get the idea of filter buckets. Could you please explain them to me again? 
I mean I get what they do, but I'm missing a good example to understand their need.

Reply:
we removed the filter bucket.